contrib/utilities/update-copyright.sh

```diff
@@ -24,51 +25,162 @@ if test ! -d source -o ! -d include -o ! -d examples ; then
+{
+  file="${1}"
+
+  if ! [ -f ${file} ]; then
+    echo "Skipping ${file}: not a file"
+    return
+  fi
+
+  if ! head -13 ${file} | grep -q "^.. This file is part of the deal.II library.$" ; then
+    echo "Skipping ${file}: no deal.II copyright header"
+    return
+  fi
 
-files="
-  $(echo contrib/*/*.{py,sh} \
-         contrib/python-bindings/CMakeLists.txt \
-         contrib/python-bindings/*/*.{h,cc,py} \
-         contrib/utilities/{update-copyright,indent} \
-         doc/doxygen/*/*.{h,h.in} \
-         doc/doxygen/scripts/*.pl \
-         doc/screen.css)
-  $(find include/ source/ examples/ | egrep '\.(h|in|cc|cu)$')
-  $(find cmake/ | egrep '\.(cmake|in|cc)$')
-  $(find . -name CMakeLists.txt)
-  $(find tests/ | egrep '\.(h|cc)$')
-  $(find doc/ | egrep '\.html$')
-"
-
-
-for i in $files ; do
-  # get the last year this file was modified from the git log.
-  # we don't want to see patches that just updated the copyright
-  # year, so output the dates and log messages of the last 3
-  # commits, throw away all that mention both the words
-  # "update" and "copyright", and take the year of the first
-  # message that remains
   #
-  # (it should be enough to look at the last 2 messages since
-  # ideally no two successive commits should have updated the
-  # copyright year. let's err on the safe side and take the last
-  # 3 commits.)
-  last_year=`git log -n 3 --date=short --format="format:%cd %s" $i | \
-             egrep -i -v "update.*copyright|copyright.*update" | \
-             head -n 1 | \
-             perl -p -e 's/^(\d\d\d\d)-.*/\1/g;'`
-
-  # get the first year this file was modified from the actual
-  # file. this may predate the git log if the file was copied
-  # from elsewhere
-  first_year=`cat $i | egrep 'Copyright \(C\) [0-9]{4}' | \
-              perl -p -e "s/.*Copyright \(C\) (\d{4}).*/\1/g;"`
-
-  # print a status message. we really only have to update
-  # the copyright year if the first and last year are
-  # different
-  echo "Processing $i: ${first_year} - ${last_year}"
-  if test ! "${first_year}" = "${last_year}" ; then
-    perl -pi -e "s/(Copyright \(C\) \d{4})( - \d{4})?(, \d{4}( - \d{4})?)*/\1 - ${last_year}/g;" $i
+  # Get the last year this file was modified from the git log. We don't
+  # want to see patches that just updated the copyright year, thus find the
+  # first commit that
+  #  - does not mention both the words "update" and "copyright", as well as
+  #  - "Update license headers".
+  #
+
+  last_year=`git log -n 3 --date=short --format="format:%cd %s" ${file} | \
+    egrep -i -v "update.*copyright|copyright.*update|Update license headers" | \
+    head -n 1 | \
+    perl -p -e 's/^(\d\d\d\d)-.*/\1/g;'`
+
+  #
+  # It should not happen, that the grep removes all 3 most recent commits
+  # simultaneously but if it does then run the git log command again with
+  # full history:
+  #
+
+  [ -z "$last_year" ] && last_year=`git log --date=short --format="format:%cd %s" ${file} | \
```

Comment:
Out of curiosity, why didn't you just write this as `if test -z "$last_year" ; then ...`?

Reply:
This is really purely a stylistic choice:
 - `test` and `[` are the same shell built-in. And `[ -z ... ]` looks better than `test -z ...`. I should have probably used the extended test pattern `[[ ... ]]` consistently...
 - I tend to use the `[[ ... ]] && VARIABLE="..."` a lot for simple conditional variable assignments. That's somewhat a bash coding style coming from Gentoo ebuilds. But yes, we could have used an `if` statement as well.

Reply:
I see. It's perhaps a bit harder to read, but I'm ok with the stylistic choice.